File: SnowForecast/app/src/main/java/com/sners/snowforecast/data/Hourly.java
package com.sners.snowforecast.data;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashSet;
import java.util.Set;


public class Hourly {

    String summary;
    String icon;
    ArrayList<IntervalData> hourlyData;

    private Float maxPrecip;

    Set<Integer> weatherCodes = new HashSet<Integer>();
    Set<Integer> precipCodes = new HashSet<Integer>();
    WeatherHelper weatherHelper = new WeatherHelper();
    final int numPointsToPlot = 48;


    ////////////////////////////////////////////////////////////////////////////////
    //	Constructor
    ////////////////////////////////////////////////////////////////////////////////
    public Hourly(JSONArray hourlyArary) {
        maxPrecip = -1.0f;
        hourlyData = new ArrayList<IntervalData>();
        try {

            summary = "Please implement me";
            icon = "clear day";
            Boolean beforeNow = true;
            Calendar rightNow = Calendar.getInstance();
            int hourNow = rightNow.get(Calendar.HOUR_OF_DAY);


            for (int intervalCounter = 0; intervalCounter < hourlyArary.length(); intervalCounter++) {
                HourlyData dataInst = new com.sners.snowforecast.data.HourlyData(hourlyArary.getJSONObject(intervalCounter));
                if (beforeNow) {
                    int hourInst = Integer.parseInt(dataInst.getTime().substring(0, 2));
                    if (hourInst >= hourNow) {
                        beforeNow = false;
                    }
                }
                if (!beforeNow) {
                    hourlyData.add(dataInst);
                }

            }

        } catch (JSONException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }

    }


    public Float getMaxPrecip() {
        if (maxPrecip < 0.0f) {
            //	only calculate it once
            for (int hourCounter = 0; hourCounter < numPointsToPlot; hourCounter++) {

                float precip = hourlyData.get(hourCounter).getPrecipIntensity();
                if (precip > maxPrecip) {
                    maxPrecip = precip;
                }
            }
        }
        return maxPrecip;
    }

    public String getSummary() {
        return summary;
    }

    public String getIcon() {
        return icon;
    }

    public ArrayList<IntervalData> getHourlyData() {
        return hourlyData;
    }


}

File: SnowForecast/app/src/main/java/com/sners/snowforecast/data/Hourly.kt
package com.sners.snowforecast.data

import com.sners.snowforecast.weather.WeatherConstants
import com.sners.snowforecast.weather.WeatherHelper
import org.json.JSONArray
import org.json.JSONException
import java.util.*

/**
 * A class for managing hourly weather conditions
 * It takes the raw json and converts it into a data structure
 * @param hourlyArray The raw current data coming in from the API call
 * @param currentDateUnix The reference date used for working out when is now
 */
class Hourly(hourlyArray: JSONArray, val currentDateUnix : Long) {

    /**
     *  @property hourlyData Array of data objects
     */
    var hourlyData: ArrayList<IntervalData>

    /**
     *  @property summary Weather summary for the hour
     */
    var summary: String? = null

    /**
     *  @property icon String indicating which icon to use
     */
    var icon: String? = null

    /**
     *  @property maxPrecip Max precipitation over all the graphable hours
     */
    private var maxPrecip: Float

    /**
     *  @property weatherHelper A class for common weather functions
     */
    private val weatherHelper: WeatherHelper

    /**
     * Get max precip
     * @return Maximimum precip for the period
     */
    fun getMaxPrecip(): Float {
        if (maxPrecip < 0.0f) {
            //	only calculate it once
            for (hourCounter in 0..WeatherConstants.HOURLY_NUM_POINTS_TO_PLOT - 1) {
                val precip = hourlyData[hourCounter].precipIntensity
                if (precip > maxPrecip) {
                    maxPrecip = precip
                }
            }
        }
        return maxPrecip
    }

    /*
  Constructor
   */
    init {
        weatherHelper = WeatherHelper()
        maxPrecip = -1.0f
        hourlyData = ArrayList()
        try {
            summary = "Please implement me"
            icon = "clear day"
            var beforeNow = true

            for (intervalCounter in 0..hourlyArray.length() - 1) {
                val dataInst = HourlyData(hourlyArray.getJSONObject(intervalCounter))
                //  This API returns hours in the past, it starts at 0am today
                //  We only want from the current hour onwards
                if (beforeNow) {

                    if (dataInst.timeUnix ?: Long.MAX_VALUE >= currentDateUnix) {
                        beforeNow = false
                    }
                }
                if (!beforeNow) {
                    hourlyData.add(dataInst)
                }
            }
        } catch (e: JSONException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }
    }

    /**
     *  @property tempFeelsLike Feels-like Temperature
     */
    val tempFeelsLike = hourlyData[0].tempFeelsLike

    /**
     *  @property humidity Humidity
     */
    val humidity: Float
        get() = hourlyData[0].humidity

    /**
     *  @property cloudCover Cloud cover as a formatted string
     */
    val cloudCoverString: String
        get() = weatherHelper.probabilityToPercent(cloudCover)

    /**
     *  @property dewPoint Dew point temperature as a number
     */
    val dewPoint = hourlyData[0].dewPoint

    /**
     *  @property cloudCover Cloud Cover
     */
    val cloudCover = hourlyData[0].cloudCover

}